File: examples/python/resolve.py
#!/usr/bin/env python3
"""
Resolve a pay: alias using the DNS of Money API.

pay: aliases are human-readable financial addresses defined by the FAS-1
specification. They resolve to payment endpoints across multiple rails
(ACH, FedNow, XRPL, SWIFT) — just like DNS resolves domain names to
IP addresses, but for payments.

    pay:vendor.alpha  →  { rail: "fednow", routing: "021000021", ... }
    pay:agent.compute →  { rail: "xrpl",   address: "rXXX...", ... }

This script shows how to resolve a pay: alias to its payment endpoints
using a simple HTTP GET request.

Requirements:
    pip install requests

Usage:
    python resolve.py
    API_KEY=your_key python resolve.py
"""

import os
import sys

import requests

BASE_URL = "https://api.dnsofmoney.com"
ALIAS = "pay:vendor.alpha"


def resolve(alias: str, api_key: str | None = None) -> dict:
    """
    Resolve a pay: alias to its payment endpoints.

    Args:
        alias: A FAS-1 pay: URI (e.g., "pay:vendor.alpha")
        api_key: Optional API key for authenticated resolution (higher tier)

    Returns:
        The full resolution response as a dict.
    """
    headers = {}
    if api_key:
        headers["X-API-Key"] = api_key

    resp = requests.get(f"{BASE_URL}/v1/resolve/{alias}", headers=headers)

    # Handle common error cases
    if resp.status_code == 404:
        print(f"Alias not found: {alias}")
        print(f"Error: {resp.json().get('detail', 'ALIAS_NOT_FOUND')}")
        sys.exit(1)

    if resp.status_code == 403:
        print(f"Resolution blocked (compliance): {resp.json().get('detail')}")
        sys.exit(1)

    resp.raise_for_status()
    return resp.json()


def main():
    api_key = os.environ.get("API_KEY")

    print(f"Resolving {ALIAS} ...")
    data = resolve(ALIAS, api_key)

    # --- Read the resolution response ---

    # Entity info: who owns this alias
    entity = data.get("entity", {})
    print(f"\nEntity:         {entity.get('display_name', 'N/A')}")
    print(f"Entity type:    {entity.get('entity_type', 'N/A')}")

    # Preferred rail: the recommended payment method
    preferred = data.get("preferred_rail", "N/A")
    print(f"Preferred rail: {preferred}")

    # ISO 20022 hint: which message family to use
    iso_hint = data.get("iso20022_hint", {})
    print(f"ISO 20022 hint: {iso_hint.get('message_type', 'N/A')}")

    # X-Resolved-From header: was this served from cache or origin?
    # (This header is on the HTTP response, not in the JSON body)
    # Access it via resp.headers["X-Resolved-From"] if using the raw response.

    # All available payment endpoints, ranked by priority
    endpoints = data.get("endpoints", [])
    print(f"\nEndpoints ({len(endpoints)}):")
    for ep in endpoints:
        rail = ep.get("rail", "unknown")
        priority = ep.get("priority", "?")
        currency = ep.get("currency", "USD")
        fee = ep.get("fee_estimate", "N/A")
        print(f"  [{priority}] {rail} ({currency}) — fee estimate: {fee}")

    # TTL: how long this resolution is valid
    ttl = data.get("ttl_seconds", 300)
    print(f"\nCache TTL: {ttl}s")
    print(f"Resolved at: {data.get('resolved_at', 'N/A')}")


if __name__ == "__main__":
    main()

File: examples/typescript/resolve.ts
/**
 * Resolve a pay: alias using the DNS of Money API.
 *
 * pay: aliases are human-readable financial addresses defined by FAS-1.
 * They resolve to payment endpoints across multiple rails (ACH, FedNow,
 * XRPL, SWIFT) — like DNS for payments.
 *
 * This example uses the built-in fetch() API — no external dependencies.
 *
 * Usage:
 *   npx tsx resolve.ts
 *   API_KEY=your_key npx tsx resolve.ts
 */

const BASE_URL = "https://api.dnsofmoney.com";
const ALIAS = "pay:vendor.alpha";

// --- Types ---

interface PaymentEndpoint {
  rail: "ach" | "fednow" | "swift" | "xrpl" | "solana" | "card";
  currency: string;
  priority: number;
  fee_estimate?: string;
  settlement_latency?: string;
  routing_metadata?: Record<string, unknown>;
}

interface Entity {
  display_name: string;
  entity_type: string;
  tier?: "founding" | "standard" | "enterprise";
  jurisdiction?: string;
}

interface ISO20022Hint {
  message_type: string;
  service_level?: string;
  local_instrument?: string;
  charge_bearer?: string;
}

interface Compliance {
  status: "clear" | "pending" | "blocked";
  screened_at?: string;
  travel_rule_required?: boolean;
}

interface ResolutionResponse {
  alias: string;
  resolved: boolean;
  entity?: Entity;
  preferred_rail?: string;
  endpoints: PaymentEndpoint[];
  iso20022_hint?: ISO20022Hint;
  compliance?: Compliance;
  ttl_seconds: number;
  resolved_at: string;
  resolution_status?: "resolved" | "partial" | "failed";
  caller_tier?: number;
  resolved_from?: "origin" | "cache";
  error_code?: string;
}

// --- Resolve ---

async function resolve(
  alias: string,
  apiKey?: string
): Promise<ResolutionResponse> {
  const headers: Record<string, string> = {};
  if (apiKey) {
    headers["X-API-Key"] = apiKey;
  }

  const resp = await fetch(`${BASE_URL}/resolve/${alias}`, { headers });

  // Handle 404 — alias not found
  if (resp.status === 404) {
    const body = await resp.json();
    throw new Error(`Alias not found: ${alias} (${body.detail ?? "ALIAS_NOT_FOUND"})`);
  }

  // Handle 403 — compliance blocked
  if (resp.status === 403) {
    const body = await resp.json();
    throw new Error(`Resolution blocked: ${body.detail ?? "COMPLIANCE_BLOCKED"}`);
  }

  if (!resp.ok) {
    throw new Error(`HTTP ${resp.status}: ${resp.statusText}`);
  }

  return resp.json();
}

// --- Main ---

async function main() {
  const apiKey = process.env.API_KEY;

  console.log(`Resolving ${ALIAS} ...`);

  try {
    const data = await resolve(ALIAS, apiKey);

    // Entity info
    const entity = data.entity;
    console.log(`\nEntity:         ${entity?.display_name ?? "N/A"}`);
    console.log(`Entity type:    ${entity?.entity_type ?? "N/A"}`);

    // Preferred rail
    console.log(`Preferred rail: ${data.preferred_rail ?? "N/A"}`);

    // ISO 20022 hint
    console.log(`ISO 20022 hint: ${data.iso20022_hint?.message_type ?? "N/A"}`);

    // Endpoints
    console.log(`\nEndpoints (${data.endpoints.length}):`);
    for (const ep of data.endpoints) {
      console.log(
        `  [${ep.priority}] ${ep.rail} (${ep.currency ?? "USD"}) — fee: ${ep.fee_estimate ?? "N/A"}`
      );
    }

    // Cache info
    console.log(`\nCache TTL:   ${data.ttl_seconds}s`);
    console.log(`Resolved at: ${data.resolved_at}`);
    console.log(`Served from: ${data.resolved_from ?? "unknown"}`);
  } catch (err) {
    console.error(`Error: ${(err as Error).message}`);
    process.exit(1);
  }
}

main();
